includes/free-trial/class-wc-calypso-bridge-free-trial-expired-plan-redirects.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+/**
+ * Class WC_Calypso_Bridge_Free_Trial_Expired_Plan_Redirects.
+ *
+ * @since   2.0.9
+ * @version 2.0.9
+ *
+ * Detects when we have an expired eCommerce trial plan, and redirects to Calypso for that case.
+ */
+class WC_Calypso_Bridge_Free_Trial_Expired_Plan_Redirects
+
+{
+	/**
+	 * The single instance of the class.
+	 *
+	 * @var object
+	 */
+	protected static $instance = null;
+
+	/**
+	 * Get class instance.
+	 *
+	 * @return object Instance.
+	 */
+	final public static function get_instance() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	public function __construct() {
+		if ( ! wc_calypso_bridge_is_ecommerce_trial_plan() ) {
+			return;
+		}
+
+		add_action( 'admin_init', array( $this, 'maybe_redirect_wp_admin_to_expired_plan_page' ) );
+	}
+
+	/**
+	 * When we are handling an incoming WP Admin UI request
+	 * AND the site has an expired eCommerce trial,
+	 * redirect to the expired trial page in Calypso.
+	 *
+	 * @return void
+	 */
+	public function maybe_redirect_wp_admin_to_expired_plan_page() {
+		if ( ! wc_calypso_bridge_is_ecommerce_trial_plan() ) {
+			return;
+		}
+
+		if ( defined( 'DOING_AJAX' ) && DOING_AJAX ) {
+			return;
+		}
+
+		if ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
+			return;
+		}
+
+		if ( defined( 'REST_API_REQUEST' ) && REST_API_REQUEST ) {
+			return;
+		}
+
+		$site_purchases = wpcom_get_site_purchases();
+
+		$trial_plan_purchases = array_filter(
+			$site_purchases,
+			function ( $site_purchase ) {
+				return 'ecommerce-trial-bundle-monthly' === ( $site_purchase['product_slug'] ?? null );
+			}
+		);
+
+		// TODO: if we don't have a trial plan purchase, should we still redirect?
+		if ( empty( $trial_plan_purchases ) ) {
```

Comment:
Given that the class will bail early in the constructor if there are no ecom features, the purchases array shouldn't be empty. If it is, for some reason, then it's probably something wrong with the system altogether, so just `return`ing here seems correct.

Reply:
I've updated the comment to reflect that this is an unexpected case where we won't take action.